compressor/gzip/gzip.go

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package gzip provides a YARPC binding for gzip compression.
+package gzip
+
+import (
+	"compress/gzip"
+	"io"
+
+	"go.uber.org/yarpc/api/transport"
+)
+
+// Compressor represents the gzip compression strategy.
+type Compressor struct{}
+
+var _ transport.Compressor = Compressor{}
+
+// Name is gzip.
+func (Compressor) Name() string {
+	return "gzip"
+}
+
+// Compress creates a gzip compressor.
+func (Compressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	return gzip.NewWriter(w), nil
+}
+
+// Decompress creates a gzip decompressor.
+func (Compressor) Decompress(r io.Reader) (io.Reader, error) {
```

Comment:
As per the [gzip docs](https://golang.org/pkg/compress/gzip/#Reader)
> Close closes the Reader. It does not close the underlying io.Reader. In order for the GZIP checksum to be verified, the reader must be fully consumed until the io.EOF.

Should the API be a `ReadCloser` here? Does gRPC guarantee that it will call `Close` if the method exists?

Reply:
gRPC intercepts reads until EOF. Streams that are not read to completion did not get thrown back to the pool. They do not appear to close their readers. I would be inclined to return a ReadCloser so transports have the option of early termination, but I’ll have to check on the interface compatibility since this doesn’t look like Haskell.

Reply:
Definitely not Haskell. https://play.golang.org/p/lW1j-HvAiUf

We’ll have to play by gRPC-Go’s rules and return io.Reader instead of io.ReadCloser, or use one type for our Compressor and register a different one for gRPC.